Nim 2.0.2 regression: macros.getType(…).kind of object type is nnkBracketExpr (was nnkObjectTy in Nim 2.0.0)

### Description

`getType` from `macros` returns an `nnkBracketExpr` for an object type with Nim 2.0.2 where Nim 2.0.0 returned an `nnkObjectTy`:

```nim
import macros

type
  Container = object
    foo: string

proc canBeImplicit(t: typedesc): string {.compileTime.} =
  ## returns empty string if type can be implicit, else the reason why it can't
  let tDesc = getType(t)
  if tDesc.kind != nnkObjectTy: return "type is not an object, but a " & $tDesc.kind
  return ""

const res = canBeImplicit(Container)
assert len(res) == 0, "unexpected error for canBeImplicit: " & res
```

### Nim Version

Nim 2.0.2 on macOS, Linux and Window, see [CI run of the test above](https://github.com/flyx/NimYAML/actions/runs/7290780284)

### Current Output

```text
/home/runner/work/NimYAML/NimYAML/test/tnimregress.nim(14) tnimregress
/home/runner/.choosenim/toolchains/nim-2.0.2/lib/std/assertions.nim(41) failedAssertImpl
/home/runner/.choosenim/toolchains/nim-2.0.2/lib/std/assertions.nim(36) raiseAssert
/home/runner/.choosenim/toolchains/nim-2.0.2/lib/system/fatal.nim(53) sysFatal
Error: unhandled exception: /home/runner/work/NimYAML/NimYAML/test/tnimregress.nim(14, 1) `len(res) == 0` unexpected error for canBeImplicit: type is not an object, but a nnkBracketExpr [AssertionDefect]
Error: execution of an external program failed: '/home/runner/work/NimYAML/NimYAML/test/tnimregress'
Error: Process completed with exit code 1.
```


### Expected Output

```text
(nothing)
```


### Possible Solution

_No response_

### Additional Information

This breaks NimYAML, specifically its [implicit](https://nimyaml.org/api/annotations.html#implicit.t) pragma. NimYAML callers that do not use this feature are not affected.

--- a/compiler/vmgen.nim
+++ b/compiler/vmgen.nim
@@ -2124,7 +2124,7 @@ proc gen(c: PCtx; n: PNode; dest: var TDest; flags: TGenFlags = {}) =
       genRdVar(c, n, dest, flags)
     of skParam:
       if s.typ.kind == tyTypeDesc:
-        genTypeLit(c, s.typ, dest)
+        genTypeLit(c, s.typ.skipTypes({tyTypeDesc}), dest)
       else:
         genRdVar(c, n, dest, flags)
     of skProc, skFunc, skConverter, skMacro, skTemplate, skMethod, skIterator:

--- a/tests/vm/ttypedesc.nim
+++ b/tests/vm/ttypedesc.nim
@@ -16,3 +16,16 @@ block: # issue #15760
 
   doAssert x[SpecialBanana]() == "SpecialBanana"
   doAssert y(SpecialBanana) == "SpecialBanana"
+
+import macros
+
+block: # issue #23112
+  type Container = object
+    foo: string
+
+  proc canBeImplicit(t: typedesc) {.compileTime.} =
+    let tDesc = getType(t)
+    doAssert tDesc.kind == nnkObjectTy
+
+  static:
+    canBeImplicit(Container)

--- a/tests/vm/tvmmisc.nim
+++ b/tests/vm/tvmmisc.nim
@@ -4,8 +4,7 @@ import os
 # bug #4462
 block:
   proc foo(t: typedesc) {.compileTime.} =
-    assert sameType(getType(t), getType(typedesc[int]))
-    assert sameType(getType(t), getType(type int))
+    assert sameType(getType(t), getType(int))
 
   static:
     foo(int)

